src/features/machine/rectMachine.js

```diff
@@ -0,0 +1,333 @@
+import Victor from 'victor'
+import Machine from './machine'
+import { distance } from '../../common/Geometry'
+
+export default class RectMachine extends Machine {
+  constructor(vertices, settings) {
+    super()
+    this.vertices = vertices
+    this.settings = settings
+    this.sizeX = Math.abs((settings.maxX - settings.minX) / 2.0)
+    this.sizeY = Math.abs((settings.maxY - settings.minY) / 2.0)
+  }
+
+  addEndpoints() {
+    if (this.settings.rectOrigin.length === 1) {
+      // OK, let's assign corners indices:
+      // [1]   [2]
+      //
+      //
+      // [0]   [3]
+      const corner = this.settings.rectOrigin[0]
+      const corners = [
+        new Victor(-this.sizeX, -this.sizeY),
+        new Victor(-this.sizeX, this.sizeY),
+        new Victor(this.sizeX, this.sizeY),
+        new Victor(this.sizeX, -this.sizeY)
+      ]
+
+      const first = this.vertices[0]
+      const last = this.vertices[this.vertices.length-1]
+      const maxRadius = Math.sqrt(Math.pow(2.0*this.sizeX, 2.0) + Math.pow(2.0*this.sizeY, 2.0)) / 2.0
+      let scale, outPoint
+
+      if (first.magnitude() <= last.magnitude()) {
+        // It's going outward
+        scale = maxRadius / last.magnitude()
+        outPoint = last
+      } else {
+        scale = maxRadius / first.magnitude()
+        outPoint = first
+      }
+
+      let clipped = this.clipLine(
+        this.vertices[this.vertices.length - 1],
+        Victor.fromObject(outPoint).multiply(new Victor(scale, scale))
+      )
+      this.vertices.push(clipped[clipped.length - 1])
+      this.vertices = [this.vertices, this.tracePerimeter(this.vertices[this.vertices.length - 1], corners[corner], true)].flat()
+    }
+
+    return this
+  }
+
+  // Returns the distance along the perimeter between two points
+  perimeterDistance(v1, v2) {
+    return this.distance(this.tracePerimeter(v1, v2, true))
+  }
+
+  // Returns whether a given path lies on the perimeter of the rectangle
+  onPerimeter(v1, v2, delta=.00001) {
+    const dx = Math.abs(Math.abs(v1.x) - this.sizeX)
+    const dy = Math.abs(Math.abs(v1.y) - this.sizeY)
+
+    if (dx < delta || dy < delta) {
+      return v1.x === v2.x || v1.y === v2.y
```

Comment:
This would still return true if dx < delta, and v1.y === v2.y... I think that's wrong. So a line from -250,100 to 50,100 wouldn't be on the perimeter.

Reply:
I updated the tests and fixed this function (I hope).